### PR TITLE
Feat/add developer guide to readme

### PR DESCRIPTION
**Pull request by molnosikristof**

### CODE REVIEW

File: e2e/README.md [hunk 1]
```diff
@@ -37,3 +37,85 @@ If you have .env file, you can use CI or TEST_PROVIDER environment variables to
...
+
+The FEF tests based on visual testing. The basic concept is to take screenshot of every status of elements or pages and make difference to a reference picture.
+
+### Structure
+
+You can find everything regarding to end to end testing in the "e2e" folder. There are two important folders located in e2e: test and utils. The "test" contains the folders of the tests of the different pages. The "utils" contains the functions for the tests.
+
+### Test
+
+Every folder in the test folder contains a .js file and a pics folder. This is a whole page test, e.g. the ./e2e/test/select-tests folder contains every tests of ux.Intellyo.com/select.
```
Comment (oroce) [inline e2e/README.md]: can you make this into a link?

File: e2e/test/buttons-tests/buttons-test.js [hunk 2]
```diff
@@ -17,15 +17,16 @@ describe('FEF buttons tests', () => {
   it('Checks the the fef buttons page title and browser compare visual regression', () => {
     browser.url('/buttons');
     assert.equal(browser.getTitle(), 'Intellyo Application Design System');
-    assert(takeScreenshotAndGetWholePageCompareResult({defaultTolerance: 5.1, ignoreComparison: false, testDirPath: __dirname}), 'Whole FEF buttons page screenshot compare to a reference picture');
+    assert(takeScreenshotAndGetWholePageCompareResult({defaultTolerance: 5.1, ignoreComparison: false, testDirPath: __dirname}),
```
Comment (oroce) [inline e2e/test/buttons-tests/buttons-test.js]: `testDirPath` can be removed i guess

File: e2e/test/tags-input-tests/tags-input-test.js [hunk 3]
```diff
@@ -68,7 +69,7 @@ describe('FEF tags input tests', () => {
     browser.setValue(TAGS_INPUT_INPUT, `${'foobar'}\n`);
```
Comment (oroce) [inline e2e/test/tags-input-tests/tags-input-test.js]: `'foobar\n'`?

File: e2e/utils/visual-regression.js [hunk 4]
```diff
@@ -27,18 +27,21 @@ module.exports.takeScreenshotAndGetWholePageCompareResult = (options) => {
...
 
+  misMatchTolerance = options.defaultTolerance;
+
   if (options.windowsTolerance
     && browser.desiredCapabilities.platform
     && browser.desiredCapabilities.platform.includes('Windows')) {
     misMatchTolerance = options.windowsTolerance;
-  } else {
-    misMatchTolerance = options.defaultTolerance;
+  } if (options.firefoxTolerance
```
Comment (oroce) [inline e2e/utils/visual-regression.js]: newline is missing

File: e2e/utils/visual-regression.js [hunk 5]
```diff
@@ -69,12 +72,16 @@ module.exports.takeScreenShotOfElement = (elementSelector, options) => {
...
```
Comment (oroce) [inline e2e/utils/visual-regression.js]: newline is missing

File: e2e/test/buttons-tests/buttons-test.js (hunk 2 above)
Comment (molnosikristof) [reply e2e/test/buttons-tests/buttons-test.js]: This is removed in Feat/add select tests PR.

File: e2e/README.md [hunk 6]
```diff
@@ -37,3 +37,85 @@ If you have .env file, you can use CI or TEST_PROVIDER environment variables to
...
+### clean the Tests
+
+If you have too many pics which is confusing you can run a command which clean every pics except of references.
+
+- `npm run clean-test`
+
+
+## Test developement guide
+
+The FEF tests based on visual testing. The basic concept is to take screenshot of every status of elements or pages and make difference to a reference picture.
```
Comment (kristof0425) [inline e2e/README.md]: I would rather write it like this:
`FeF tests are based on visual testing. The basic concept is to take screenshots of every status of every element or page first and compare the difference between the reference pictures and the actual screenshots.`

File: e2e/README.md [hunk 7]
```diff
@@ -37,3 +37,85 @@ If you have .env file, you can use CI or TEST_PROVIDER environment variables to
...
+
+### Structure
+
+You can find everything regarding to end to end testing in the "e2e" folder. There are two important folders located in e2e: test and utils. The "test" contains the folders of the tests of the different pages. The "utils" contains the functions for the tests.
+
+### Test
+
+Every folder in the test folder contains a .js file and a pics folder. This is a whole page test, e.g. the [./e2e/test/select-tests](/e2e/test/select-tests/) folder contains every tests of [ux.Intellyo.com/select](http://ux.Intellyo.com/select).
```
Comment (kristof0425) [inline e2e/README.md]: ~~This is a whole page test,~~ ➡️  `These .js files test the whole page, e.g.`

File: e2e/README.md [hunk 8]
```diff
@@ -37,3 +37,85 @@ If you have .env file, you can use CI or TEST_PROVIDER environment variables to
...
+
+The tests located in the javascript file in groups.
```
Comment (kristof0425) [inline e2e/README.md]: I'm sorry, but I'm not sure what you meant here. 😞 
Maybe? `Each .js file represents a whole page test and each folder represents a page.` ?

File: e2e/README.md [hunk 9]
```diff
@@ -37,3 +37,85 @@ If you have .env file, you can use CI or TEST_PROVIDER environment variables to
...
+
+The references, screenshots and the diff pictures (created by the js file) located in the pics folder.
```
Comment (kristof0425) [inline e2e/README.md]: ~~The references, screenshots and the diff pictures (created by the js file) located in the pics folder.~~ 
⬇️
The **reference pictures**, screenshots and the diff pictures (created by the js file) **are**  located in the pics folder.

File: e2e/README.md [hunk 10]
```diff
@@ -37,3 +37,85 @@ If you have .env file, you can use CI or TEST_PROVIDER environment variables to
...
+
+The structure and the name of the pics are come from the structure of the test javascript file and the name of the test. For example the "this is the test set" describe contains a "I am the name of the test" test. This is stored in the "this-is-the-test-set" folder which contains a "I-am-the-name-of-the-test" folder (and other folders too). The pic of the element name generated from the last word of the test name, the time of the creation and the parameters of the the environment.
```
Comment (kristof0425) [inline e2e/README.md]: ...the pics ~~are~~ come from the...

File: e2e/README.md [hunk 11]
```diff
@@ -37,3 +37,85 @@ If you have .env file, you can use CI or TEST_PROVIDER environment variables to
...
+
+ An example of a reference:
+  - e2e/test/buttons-tests/pics/reference/fef-buttons-tests/should-check-the-button:-yeah-button/yeah-button_reference_pic.png
+
+  An example of a screenshot:
+  - e2e/test/buttons-tests/pics/screen/fef-buttons-tests/should-check-the-button:-save-button/05-03-20-pm-2018-01-14_element_windows 10_element_chrome_v63_1400x982.png
+
+### Utils
+
+As it is mentioned above : The utils contains the functions for the tests. You can find here the tools for testing.
```
Comment (kristof0425) [inline e2e/README.md]: ...The utils contain~~s~~ the...

File: e2e/README.md [hunk 12]
```diff
@@ -37,3 +37,85 @@ If you have .env file, you can use CI or TEST_PROVIDER environment variables to
...
+
+The most important two function are the takeScreenshotAndGetWholePageCompareResult and takeScreenShotOfElement. The first takes picture of a whole page and compare to the reference. The second takes a screenshot of an element and make difference to the reference.
```
Comment (kristof0425) [inline e2e/README.md]: 1.  ~~The most  important two function~~ ➡️  The **two** most important function**s** are...

2. ~~and compare to the reference.~~ ➡️  ...and compare **it** to the reference.

3. ~~and make difference~~ ➡️  ...create a diff picture by comparing it to the reference picture.

File: e2e/README.md [hunk 13]
```diff
@@ -37,3 +37,85 @@ If you have .env file, you can use CI or TEST_PROVIDER environment variables to
...
+
+### Ok... Ok... How can I write a test?
+
+Ok let's write a test. Create a folder e.g. ./e2e/test/accordions-tests. Create a test file called accordion-test.js. Copy the structure of the tests from another "something-test.js". Rename the "describe" which will be your test set name. Add an "it" and gave a name, this will be your test. First of all write in the it:
```
Comment (kristof0425) [inline e2e/README.md]: You can wrap a the folder tree into a pair of ` backticks.

File: e2e/README.md [hunk 14]
```diff
@@ -37,3 +37,85 @@ If you have .env file, you can use CI or TEST_PROVIDER environment variables to
...
+```
```
Comment (kristof0425) [inline e2e/README.md]: You can tell Github what kind of a file is put in the code block. by writing this:
```md
```js
```

**Comment by molnosikristof** [comment]
Thanks for the comments, all of them was useful.

File: e2e/README.md [hunk 15]
```diff
@@ -37,3 +37,90 @@ If you have .env file, you can use CI or TEST_PROVIDER environment variables to
...
+
+FeF tests are based on visual testing. The basic concept is to take screenshots of every status of every element or page first and compare the difference between the reference pictures and the actual screenshots.
+
+### Structure
+
+You can find everything regarding to end to end testing in the "e2e" folder. There are two important folders located in e2e: test and utils. The "test" contains the folders of the tests of the different pages. The "utils" contains the functions for the tests.
+
+### Test
+
+Every folder in the test folder contains a .js file and a pics folder. These .js files test the whole page, e.g. the [./e2e/test/select-tests](/e2e/test/select-tests/) folder contains every tests of [ux.Intellyo.com/select](http://ux.Intellyo.com/select).
```
Comment (oroce) [inline e2e/README.md]: can this `Intellyo` be `intellyo`?

File: e2e/README.md [hunk 16]
```diff
@@ -37,3 +37,90 @@ If you have .env file, you can use CI or TEST_PROVIDER environment variables to
...
+
+Each .js file represents a whole page test and each folder represents a page.
+
+The reference pictures, screenshots and the diff pictures (created by the js file) located in the pics folder.
+
+The structure and the name of the pics come from the structure of the test javascript file and the name of the test. For example the "this is the test set" describe contains a "I am the name of the test" test. This is stored in the "this-is-the-test-set" folder which contains a "I-am-the-name-of-the-test" folder (and other folders too). The pic of the element name generated from the last word of the test name, the time of the creation and the parameters of the the environment.
```
Comment (oroce) [inline e2e/README.md]: wont this `I-am-the-name-of-the-test` rather be `i-am-the-name-of-the-test`?

File: e2e/README.md [hunk 17]
```diff
@@ -37,3 +37,90 @@ If you have .env file, you can use CI or TEST_PROVIDER environment variables to
...
+const { takeScreenshotAndGetWholePageCompareResult, takeScreenShotOfElement } = visualRegression;
+
+describe('Your test set name', () => {
+
+  it('your test name', () => {
+    browser.url('/accordions');
+     [Do something here]
+    assert(takeScreenShotOfElement(ELEMENT_SELECTOR,
+    {windowsTolerance: 12, firefoxTolerance: 15, defaultTolerance: 5.5, ignoreComparison: false}),
+    'print it out if the test fails);
```
Comment (oroce) [inline e2e/README.md]: missing the closing single quote

File: e2e/README.md [hunk 18]
```diff
@@ -37,3 +37,90 @@ If you have .env file, you can use CI or TEST_PROVIDER environment variables to
...
+
+### pictures
+
+The results of the tests are written in the terminal, but you can check the created pictures about the elements in the same folder where the tests are located in the ./pics folder, next to the references.
+
+### run only one test file
+
+If you want to run just one file you can use the --spec argument. This argument not works with `npm run test:e2e`, you have to use `./node_modules/.bin/wdio`. Here is an example to run just the buttuns-test.js:
+
+- `./node_modules/.bin/wdio wdio.conf.js --spec ./e2e/test/buttons-tests/buttons-test.js`
```
Comment (oroce) [inline e2e/README.md]: wrong! `npm run test:e2e -- --spec ./e2e/test/buttons-tests/buttons-test.js` will work. more info about the double dash magc: https://docs.npmjs.com/cli/run-script

File: e2e/README.md [hunk 19]
```diff
@@ -37,3 +37,90 @@ If you have .env file, you can use CI or TEST_PROVIDER environment variables to
...
+
+FeF tests are based on visual testing. The basic concept is to take screenshots of every status of every element or page first and compare the difference between the reference pictures and the actual screenshots.
+
+### Structure
+
+You can find everything regarding to end to end testing in the "e2e" folder. There are two important folders located in e2e: test and utils. The "test" contains the folders of the tests of the different pages. The "utils" contains the functions for the tests.
+
+### Test
+
+Every folder in the test folder contains a .js file and a pics folder. These .js files test the whole page, e.g. the [./e2e/test/select-tests](/e2e/test/select-tests/) folder contains every tests of [ux.Intellyo.com/select](http://ux.intellyo.com/select).
```
Comment (oroce) [inline e2e/README.md]: still `intellyo` and not `Intellyo`:D

**Pull request merged**